src/memo/memo.ml

```diff
@@ -685,254 +638,413 @@ let create_hidden (type output) name ?doc ~input typ impl =
+  end
+
+  let do_validate (dep_node : _ Dep_node.t) inp ivar running_state =
     let* res =
       Call_stack.push_async_frame
         (T { without_state = dep_node.without_state; running_state })
         (fun () ->
-          match dep_node.without_state.spec.f with
-          | Function.Async f ->
-            (* A consequence of using [Fiber.collect_errors] is that memoized
-               functions don't report errors promptly - errors are reported once
-               all child fibers terminate. To fix this, we should use
-               [Fiber.with_error_handler], but we dont have access to dune's
-               error reoprting mechanism in memo *)
-            Fiber.collect_errors (fun () -> f inp))
-    in
-    (* update the output cache with the correct value *)
-    let deps = List.rev running_state.deps_so_far.deps_reversed in
-    let res =
-      Result.map_error res ~f:(fun exns ->
-          (* this step deduplicates the errors *)
-          Value.Async (Exn_set.of_list exns))
+          let* from_cache =
+            match dep_node.last_cached_value with
+            | None ->
+              Fiber.return
+                (Prev_cycle_cache_lookup_result.Invalid { old_value = None })
+            | Some cv -> (
+              let+ res = deps_changed cv.deps in
+              match res with
+              | Unchanged ->
+                dep_node.last_cached_value <-
+                  Some { cv with last_validated_at = Run.current () };
+                Prev_cycle_cache_lookup_result.Valid cv
+              | Changed ->
+                dep_node.last_cached_value <- None;
+                Invalid { old_value = Some cv } )
+          in
+          match from_cache with
+          | Valid v -> Fiber.return v
+          | Invalid { old_value } -> (
+            match dep_node.without_state.spec.f with
+            | Function.Async f ->
+              (* A consequence of using [Fiber.collect_errors] is that memoized
+                 functions don't report errors promptly - errors are reported
+                 once all child fibers terminate. To fix this, we should use
+                 [Fiber.with_error_handler], but we dont have access to dune's
+                 error reoprting mechanism in memo *)
+              let+ res = Fiber.collect_errors (fun () -> f inp) in
+              let res =
+                Result.map_error res ~f:(fun exns ->
+                    (* this step deduplicates the errors *)
+                    Value.Async (Exn_set.of_list exns))
+              in
+              (* update the output cache with the correct value *)
+              let deps_rev = running_state.deps_so_far.deps_reversed in
+              let res =
+                match old_value with
+                | None -> Cached_value.create res ~deps_rev
+                | Some old_cv -> (
+                  match Cached_value.dep_changed dep_node old_cv.value res with
+                  | false -> Cached_value.confirm_old_value ~deps_rev old_cv
```

Comment:
Same concern: maybe we should mutate the old record instead of creating a new one?

Reply:
I changed it to do the mutation, but, as above, it's not a correctness issue.